## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.

application_df = application_df.drop( ['EIN','NAME'], axis=1)



In [3]:
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [4]:
# Determine the number of unique values in each column.

application_df_unique = application_df.nunique().sort_values()
print(application_df_unique)

STATUS                       2
SPECIAL_CONSIDERATIONS       2
IS_SUCCESSFUL                2
ORGANIZATION                 4
USE_CASE                     5
AFFILIATION                  6
INCOME_AMT                   9
APPLICATION_TYPE            17
CLASSIFICATION              71
ASK_AMT                   8747
dtype: int64


In [5]:
# Look at APPLICATION_TYPE value counts for binning
#  YOUR CODE GOES HERE
application_df['APPLICATION_TYPE'].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Choose a cutoff value
# create a list of application types to be replaced
# use the variable name `application_types_to_replace`

APPLICATION_TYPE_CUTOFF = 500

application_type_replace_filter_series = application_df["APPLICATION_TYPE"].value_counts() < APPLICATION_TYPE_CUTOFF
application_type_replace_filter_series


T3     False
T4     False
T6     False
T5     False
T19    False
T8     False
T7     False
T10    False
T9      True
T13     True
T12     True
T2      True
T25     True
T14     True
T29     True
T15     True
T17     True
Name: APPLICATION_TYPE, dtype: bool

In [7]:
application_types_to_replace = application_df["APPLICATION_TYPE"].value_counts()[application_type_replace_filter_series].index

In [8]:
# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [9]:
application_df.loc[25:50]

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
25               T4       Independent          C3000    ProductDev   
26               T3  CompanySponsored          C1000  Preservation   
27               T3  CompanySponsored          C2000  Preservation   
28               T3  CompanySponsored          C2000  Preservation   
29               T6       Independent          C2000    ProductDev   
30               T6       Independent          C2000    ProductDev   
31               T3       Independent          C1200  Preservation   
32               T7       Independent          C1000    ProductDev   
33               T3  CompanySponsored          C1000  Preservation   
34               T3       Independent          C1000  Preservation   
35               T3       Independent          C1200  Preservation   
36               T4       Independent          C1000    ProductDev   
37               T3       Independent          C1000  Preservation   
38               T7       Independent          C1000    ProductDev   
39               T3       Independent          C1000  Preservation   
40               T3       Independent          C1000  Preservation   
41            Other       Independent          C1000    ProductDev   
42               T7       Independent          C1000    ProductDev   
43               T3  CompanySponsored          C1000  Preservation   
44               T3       Independent          C7200  Preservation   
45               T3       Independent          C1000  Preservation   
46               T3  CompanySponsored          C2000  Preservation   
47               T7       Independent          C1000    ProductDev   
48               T3       Independent          C1700  Preservation   
49               T3       Independent          C1000  Preservation   
50               T3       Independent          C1000  Preservation   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
25        Trust       1              0                      N     5000   
26        Trust       1              0                      N     5000   
27        Trust       1    25000-99999                      N    44004   
28        Trust       1              0                      N     5000   
29        Trust       1              0                      N     5000   
30        Trust       1  100000-499999                      N   323466   
31        Trust       1              0                      N     5000   
32        Trust       1              0                      N     5000   
33        Trust       1              0                      N     5000   
34        Trust       1              0                      N     5000   
35        Trust       1              0                      N     5000   
36        Trust       1              0                      N     5000   
37        Trust       1              0                      N     5000   
38  Association       1              0                      N     5000   
39        Trust       1    25000-99999                      N     5000   
40        Trust       1    25000-99999                      N   709815   
41        Trust       1    25000-99999                      N  1329410   
42        Trust       1              0                      N     5000   
43        Trust       1              0                      N     5000   
44        Trust       1              0                      N     5000   
45        Trust       1              0                      N     5000   
46        Trust       1              0                      N     5000   
47        Trust       1              0                      N     5000   
48        Trust       1              0                      N     5000   
49        Trust       1              0                      N     5000   
50        Trust       1  100000-499999                      N    31613   

    IS_SUCCESSFUL  
25              1  
26              1  
27              1  
28              0  
29           

In [10]:
# Look at CLASSIFICATION value counts for binning

application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [11]:
# You may find it helpful to look at CLASSIFICATION value counts >1
classification_gt_1_filter_series = application_df['CLASSIFICATION'].value_counts()>1
application_df['CLASSIFICATION'].value_counts()[classification_gt_1_filter_series]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [12]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`

CLASSIFICATION_TYPE_CUTOFF = 100

classification_replace_filter_series = application_df["CLASSIFICATION"].value_counts() < CLASSIFICATION_TYPE_CUTOFF
classification_replace_filter_series
classifications_to_replace= application_df["CLASSIFICATION"].value_counts()[classification_replace_filter_series].index


# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
Other      669
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: CLASSIFICATION, dtype: int64

In [13]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df_dummies = pd.get_dummies(application_df)

In [14]:
# Split our preprocessed data into our features and target arrays
y = application_df_dummies["IS_SUCCESSFUL"].values
# X = application_df_dummies.drop(["IS_SUCCESSFUL"],1).values
X = application_df_dummies.drop('IS_SUCCESSFUL', axis=1)



In [15]:
# Split the preprocessed data into a training and testing dataset

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [16]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [18]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE
number_input_features = len(X_train_scaled[1])
hidden_nodes_layer1 = 7
hidden_nodes_layer2 = 14
hidden_nodes_layer3 = 21

nn = tf.keras.models.Sequential()

# First hidden layer

#nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, activation="tanh", input_dim= number_input_features))
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, activation="relu", input_dim= number_input_features))
# Second hidden layer

#nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="tanh"))
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))
# Third hidden layer

#nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="tanh"))
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="relu"))

# Output layer

nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_2 (Dense)             (None, 7)                 350       
                                                                 
 dense_3 (Dense)             (None, 14)                112       
                                                                 
 dense_4 (Dense)             (None, 21)                315       
                                                                 
 dense_5 (Dense)             (None, 1)                 22        
                                                                 
Total params: 799 (3.12 KB)
Trainable params: 799 (3.12 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [19]:
# Compile the model

nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [20]:
# Train the model

fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
804/804 [==============================] - 3s 2ms/step - loss: 0.5905 - accuracy: 0.7110
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5581 - accuracy: 0.7286
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5539 - accuracy: 0.7301
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5528 - accuracy: 0.7299
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5511 - accuracy: 0.7306
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5504 - accuracy: 0.7317
Epoch 7/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5493 - accuracy: 0.7319
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5486 - accuracy: 0.7327
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5482 - accuracy: 0.7317
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5478 - accura

In [21]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5530 - accuracy: 0.7233 - 492ms/epoch - 2ms/step
Loss: 0.5530050992965698, Accuracy: 0.7232652902603149


In [22]:
# Export our model to HDF5 file
nn.save('AlphabetSoupCharity2.h5')

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


Attemp Number 2 To Optimise the Model To Acheive Accuracy Higher than 75%

For Optimising the Model in this attempt number of bins has been altered for Applitication Type column and Classification.

Number of Hidden Layers = 3

Number of Neurons for each Layer

hidden_nodes_layer1 = 7

hidden_nodes_layer2 = 14

hidden_nodes_layer3 = 21

Model has been tried and optimised using two activation fuctions :

Tanh and Relu

Result:

Making above changes to the model displays the accuracy of 72%.